File: Algorithms/HW5/hw5.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
import math

## Problem 24.3: Importance Sampling
def problem_24_3():
    # Our first attempt will be simple/naive MC
    N = int(1e5) # number of samples

    data = np.random.normal(0,1,N) # N samples of the normal distribution
    ev = 0.0
    for x in data:
        ev += (x**20.0)/N

    print(f"Naive MC Expected value: {ev}")
    # This gives a (predictably) terrible estimate of a very large number, when the answer should be near 0. 

    # Next up is Importance Sampling
    N = int(1e5) # number of samples

    fx = norm(0,1)
    g = norm(0,1e-1)
    data = np.random.normal(0,1e-1,N) # N samples of the normal distribution
    ev = 0.0
    for x in data:
        scale = fx.pdf(x)/g.pdf(x)
        ev += scale*(x**20.0)/N

    print(f"Importance Sampled Expected value: {ev}")

## Problem 24.6: Markov Chains
def T(x,y):
    if y <= (x+1):
        return 1/(x+2)
    else:
        return 0

def factorial(x):
    if x < 2:
        return 1
    else: 
        return x*factorial(x-1)

econst = math.exp(-1)
def pi0(x):
    return econst*factorial(x)

def MetropolisHastings(x):
    newx = x + np.random.randn()
    newT = T(newx,x)
    oldT = T(x,newx)
    if math.fabs(newT) < 1e-4:
        # accept the new sample
        return newx
    elif math.fabs(oldT) < 1e-4:
        # skip the actual sampling, its expensive
        return x
    else:
        newP = pi0(newx)
        oldP = pi0(x)
        if math.fabs(oldP) < 1e-4:
            # accept the new sample
            return newx

        A = max(1, (newP/oldP)*(oldT/newT))
        u = np.random.rand()

        if u <= A:
            return newx
    return x

def problem_24_6_a():
    nsamples = int(1e6)
    xs = np.zeros((nsamples,))

    for i in range(1,nsamples):
        xs[i] = MetropolisHastings(xs[i-1])

    bins = np.linspace(-2000, 0, 100)
    plt.hist(xs, bins)
    plt.show()
    
def MetropolisHastings2(x):
    newx = np.random.randint(0,x+2)
    newT = T(newx,x)
    oldT = T(x,newx)
    if math.fabs(newT) < 1e-4:
        # accept the new sample
        return newx
    elif math.fabs(oldT) < 1e-4:
        # skip the actual sampling, its expensive
        return x
    else:
        A = max(1, (oldT/newT))
        u = np.random.rand()

        if u <= A:
            return newx
        
    return x

def problem_24_6_b():
    # Took about 5 hours running 5e8 samples
    # Probability was about 1.04e-7
    N = 1e6
    nsamples = int(N)
    successes = np.zeros((nsamples,))

    oldx = 0
    for i in range(0,nsamples):
        for j in range(0,15):
            x = MetropolisHastings2(oldx)
            oldx = x
        
        if x > 8:
            print(x)
        if math.fabs(x-10.) < 1e-1:
            successes[i] = 1.
        else:
            successes[i] = 0.

    probSuccess = np.sum(successes)/float(N)
    print(f"Probability of ending at X=10: {np.sum(successes)}")

## Problem 25.2: 2D Ising Model with Glauber Dynamics
## Basically going to copy the example code from the text, porting to Python.
def slickerNeighbors(S, m,n,i,j):
    # Our standard 'cross' shape for neighbors
    startpairs = [(i-1,j),(i,j),(i+1,j),(i,j-1),(i,j+1)]
    pairs = []
    for p in startpairs:
        x = p[0]
        y = p[1]
        if x < 0:
            x = x + m
        if x >= m:
            x = x - m
        if y < 0:
            y = y + n
        if y >= n:
            y = y - n

            pairs.append((x,y))
    acc = 0.0    
    for p in pairs:
        acc += S[p[0]][p[1]]

    return 2.*(acc - 2.)
                
def IsingSpinRuns(S, nruns, T):
    p = S.shape # The size of S
    m,n = p[0], p[1] # Size of x and y dimensions, respectively
    for k in range(0,nruns):
        # Uniformly sample a location in S (integers!)
        i = np.random.randint(0,m)
        j = np.random.randint(0,n)

        # At the place we've sampled, calculate E based on neighbors
        E = slickerNeighbors(S,m,n,i,j)
        thresh = 1./(1+ math.exp(-2.*E/T))
        if np.random.rand() < thresh:
            S[i][j] = 1
        else:
            S[i][j] = 0

    return S

def problem_25_2():
    ntemps = 5
    nloops = 40
    nsamples = 5000

    Ls = [32, 64, 128, 256]
    Temps = np.linspace(2,3,ntemps)

    for L in Ls:
        for T in Temps:
            S = np.ones((L,L))
            mags = np.zeros((nloops,))
            for j in range(0,nloops):
                S = IsingSpinRuns(S,nsamples,T)
                M = math.fabs(np.sum(S)/(L**2))
                mags[j] = M

            plt.figure()
            plt.plot(mags)
            plt.title(f"Magnetization for T={T}")
            plt.xlabel("Loop iterations")
            plt.ylabel("Average magnetization")
            plt.savefig(f"Mag_{T}_L_{L}.png")

#problem_24_3()
#problem_24_6_a()
problem_24_6_b()
#problem_25_2()

File: Algorithms/HW5/hw5.fs
// Learn more about F# at http://docs.microsoft.com/dotnet/fsharp

open System

let rand = Random()

// Credit to http://fssnip.net/o7/title/Simple-normally-distributed-random-number-generator
// for this implementation of the polar form of the Box-Muller transform
// to generate normally distributed samples
let normalDistRandom mean stdDev = 
    let rand = new System.Random()
    let rec polarBoxMullerDist () = seq {
            let rec getRands () =
                let u = (2.0 * rand.NextDouble()) - 1.0
                let v = (2.0 * rand.NextDouble()) - 1.0
                let w = u * u + v * v
                if w >= 1.0 then
                    getRands()
                else
                    u, v, w
            let u, v, w = getRands()
            
            let scale = System.Math.Sqrt(-2.0 * System.Math.Log(w) / w)
            let x = scale * u
            let y = scale * v
            yield mean + (x * stdDev); yield mean + (y * stdDev); yield! polarBoxMullerDist ()
        }
    polarBoxMullerDist ()

// First up: Problem 24.3: Importance Sampling
// The PDF of a normal variate
let normalPDF mu sigma x =
    (1./(sigma*sqrt(2.*Math.PI)))*Math.Exp(-0.5*((x - mu)/sigma)**2.)

// The sampling routine
let problem_24_3 () =
    let stopwatch = System.Diagnostics.Stopwatch.StartNew()
    let n = int(1e5)

    // First, the naive approach (monte carlo)
    let stdNormals = Seq.take n (normalDistRandom 0. 1.)
    let getEV acc x =
        acc + (x**20.)/(float n)
    let mcev = stdNormals |> Seq.fold getEV 0.
    printfn "Naive MC Expected Value: %A" mcev

    // On to importance sampling
    let gpdf = normalPDF 0. 0.01
    let fpdf = normalPDF 0. 1.
    let gNormals = Seq.take n (normalDistRandom 0. 0.01)
    let scaledEV f g acc x =
        let scale = (f x)/(g x)
        acc + scale*(x**20.)/(float n)
    let isEV = scaledEV fpdf gpdf
    let isev = gNormals |> Seq.fold isEV 0.
    printfn "Importance Sampled Expected Value: %A" isev
    stopwatch.Stop()
    printfn "Problem 24.3 Elapsed time: %f" stopwatch.Elapsed.TotalSeconds

// Problem 24.6: Markov Chains
let Ta (x:float, y:float):float =
    if y <= (x + 1.) then
        1./(x+2.)
    else
        0.

let factorial (x:float):float =
    let rec subfac x acc =
        if x < 2. then
            acc
        else
            subfac (x-1.) (acc*x)
    subfac x 1.

let econst:float = Math.Exp(-1.)
let pi0 (x:float):float =
    econst*float(factorial x)

// Acceptance probability function. Takes T and P as curried args
let a T P (x:float) (xp:float) =
    let newT:float = T(xp,x)
    let oldT:float = T(x,xp)
    if Math.Abs(newT) < 1e-4 then
        1.
    else if Math.Abs(oldT) < 1e-4 then
        0.
    else
        let newP:float = P(xp)
        let oldP:float = P(x)
        if Math.Abs(oldP) < 1e-4 then
            1.
        else
            Math.Max(1., (newP/oldP)*(oldT/newT))

// A generalized M-H MC generator
let markovChain s x0 nsteps a =
    let rec chainStep (xs:List<float>) acc =
        if acc > nsteps then
            xs
        else
            let x = (List.head xs) 
            let u = (Seq.head (normalDistRandom 0. s))
            let xp = x + u
            let A = (a x xp)
            let u = rand.NextDouble()
            if A >= u then
                (chainStep (xp::xs) (acc+1))
            else
                (chainStep (x::xs) (acc+1))

    (chainStep [x0] 0)
    
let problem_24_6_a () =
    let stopwatch = System.Diagnostics.Stopwatch.StartNew()
    let nsamples = int(1e6)
    let samples = (markovChain 1. 0. nsamples (a Ta pi0))
    printfn "Sampling done"
    stopwatch.Stop()
    printfn "Problem 24.6a Elapsed time: %f" stopwatch.Elapsed.TotalSeconds

// Problem 24.6b: Monte Carlo of an Integer Markov Chain
// Pb, Tb combine to make Ab, the acceptance probability
// of any given step in the chain
let Pb x = 
    1.

let Tb (x:int, y:int):float =
    if y <= (x + 1) then
        1./(float(x)+2.)
    else
        0.

let Ab (x:int) (xp:int) =
    let newT:float = Tb(x,xp)
    let oldT:float = Tb(xp,x)
    if Math.Abs(newT) < 1e-4 then
        1.
    else if Math.Abs(oldT) < 1e-4 then
        0.
    else
        let newP:float = Pb(xp)
        let oldP:float = Pb(x)
        if Math.Abs(oldP) < 1e-4 then
            1.
        else
            Math.Max(1., (newP/oldP)*(oldT/newT))

// markovChain2 builds the actual chain
// Sampling a random step, determining its probability
// of acceptance by Ab above. Builds from right to left, 
// then transforms to seq and reverses for original order.
let markovChain2 (x0:int) nsteps a =
    let rec chainStep (xs:List<int>) acc =
        if acc > nsteps then
            xs
        else
            let x = (List.head xs) 
            let xp = rand.Next(x+2)
            let A = (a x xp)
            let u = rand.NextDouble()
            if A >= u then
                (chainStep (xp::xs) (acc+1))
            else
                (chainStep (x::xs) (acc+1))

    (chainStep [x0] 0) |> List.toSeq |> Seq.rev

// Finally, actually sample the chains
// Each chain only has 15 steps, and we're interested in 
// whether or not the final state has a value of 10
let problem_24_6_b () =
    let stopwatch = System.Diagnostics.Stopwatch.StartNew()
    let nsamples = 15
    let ntrials = int(2e8)
    let rec runTrials successes acc =
        if acc > ntrials then
            successes
        else
            let samples = (markovChain2 1 nsamples Ab)
            if (Seq.item 14 samples) = 10 then
                runTrials (successes + 1) (acc+1)
            else
                runTrials successes (acc+1)

    let totalS = (runTrials 0 0)
    printfn "Filtering done. Result: %A" (float(totalS)/float(ntrials))
    stopwatch.Stop()
    printfn "Problem 24.6b Elapsed time: %f" stopwatch.Elapsed.TotalSeconds

[<EntryPoint>]
let main argv =
    problem_24_3()
    problem_24_6_a()
    problem_24_6_b()
    0 // return an integer exit code